samples: fix the metairq_dispatch sample sporadically failed on ehl_crb

enjiamai opened the pull request:
The metairq sample sometimes failed due to stuck because it has a
potential race when recording the latencies. The previous solution
checks whether the stats.mirq greater than MAX_EVENT. But stats.mirq
might also have a race here. So we change to check the msg sequence,
its correctness ensured by the kernel's message queue.

And also remove the spinlock in previous fix because tests only use
1 CPU, no need to add this. It also improves the fix for #40889,
to prevent test failed sporadically on acrn_ehl_crb and ehl_crb.	

Fix #42874.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

andyross commented on samples/kernel/metairq_dispatch/src/main.c:
```diff
@@ -191,7 +197,7 @@ static void thread_fn(void *p1, void *p2, void *p3)
 		 */
 		while (k_cycle_get_32() - start < m.proc_cyc) {
-			for (volatile int i = 0; i < 100; i++) {
+			for (volatile int i = 0; i < SPIN_TIMES; i++) {
```
This doesn't increase the time spent waiting, it just decreases the frequency with which the timer is polled.  We wait for `m.proc_cyc` in either case.

If this is fixing something, the root cause has to be elsewhere.

enjiamai replied:
You are absolutely right. This root cause is due to race on record_latencies subroutine, add timeout here is not a right way to go. Correct it. thanks for pointing this out.

andyross commented on samples/kernel/metairq_dispatch/src/main.c:
```diff
@@ -120,13 +118,6 @@ static void record_latencies(struct msg *m, uint32_t latency)
 		return;
 	}
 
-	k_spinlock_key_t key = k_spin_lock(&lock);
```
I'm confused here.  I'm just now noticing the earlier patch, but... it looks correct to me.  This is a test of preemptible thread behavior, so locking of some form between the threads is probably what we want.

andyross replied:
But to be clear: it doesn't have to be a spinlock.  There's no ISR context and (as you point out) we're running with MP_NUM_CPUS=1, so anything will do: a k_mutex() or k_sem() would be fine, and (as long as we're never SMP) so would sched_un/lock().

enjiamai replied:
Yes, @andyross , in previous patch I did not aware of this is a 1 CPU scenario, like you pointing out, it only does irq_lock here, and no ISR context. (Actually I tried it doesn't work at all to fix this issue).  So the spinlock should be removed here.

And I try on another way to fix this, by just checking if the message sequence number > MAX_EVENT) when entering the record_latencies subroutine  (in previously I check the stats.mirq but it still failed sometimes due to this shared variable can also have a race. 

Look like it works well and no need to protect a section of code. And I tested it on both ehl_crb and acrn_ehl_crb then the test pass more than 20 times. Actually this failure can be reproduced within 5~8 times previously.

Do you think is this solution ok? Thank you!

andyross reviewed (changes requested):
I'm a little suspicious here.  Lacking a root cause, I'd say that the earlier fix (synchronize the record_latencies() function) was correct.  The new one is very confusing, and IMHO probably points to something like a latency bug with the timer driver.

enjiamai committed:
samples: fix the metairq_dispatch sample sporadically failed on ehl_crb

The metairq sample sometimes failed due to stuck because it has a
potential race when recording the latencies. The previous solution
checks whether the stats.mirq greater than MAX_EVENT. But stats.mirq
might also have a race here. So we change to check the msg sequence,
its correctness ensured by the kernel's message queue.

And also remove the spinlock in previous fix because tests only use
1 CPU, no need to add this. It also improves the fix for #40889,
to prevent test failed sporadically on acrn_ehl_crb and ehl_crb.

Fix #42874.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

andyross commented on samples/kernel/metairq_dispatch/src/main.c:
```diff
@@ -120,10 +118,11 @@ static void record_latencies(struct msg *m, uint32_t latency)
+	/* It might be a potential race condition in this subroutine.
+	 * We check if the msg seqence is reaching the MAX EVENT first.
+	 * To prevent the coming incorrect changes of the array.
+	 */
```
I'd still prefer to see a root cause here, and don't understand why the previous fix with the spinlock wasn't correct.  But this looks safe as I read it.
